ostd/src/console.rs

```diff
@@ -5,23 +5,38 @@
 use core::fmt::Arguments;
 
 /// Prints formatted arguments to the console.
-pub fn print(args: Arguments) {
-    crate::arch::console::print(args);
+///
+/// The message printed by this function will exclusively be printed without
+/// being mixed with other messages. However two distinct executions of this
+/// function on the same core may not be contiguous.
+///
+/// IRQs are disabled while printing. So do not print long messages.
+pub fn early_print(args: Arguments) {
+    use crate::sync::SpinLock;
+    static MESSAGE_LOCK: SpinLock<()> = SpinLock::new(());
+    let _lock = MESSAGE_LOCK.lock_irq_disabled();
+    crate::arch::serial::print(args);
```

Comment:
Printing messages atomically is definitely a useful feature. But I doubt the wisdom of doing this at the `early_print` function, which seems to be too low level. I think a more proper place is the logger module of OSTD.

```rust
impl log::Log for Logger {
    fn log(&self, record: &Record) { ... }
}
```

One would naturally assume the `log` method prints each record atomically. On the other hand, asking this property for `early_print` seems to be unreasonable.

In addition, I am thinking of a bigger plan for the logger mechanism. I think the OSTD's current approach to logging is inflexible. It should allow the user of OSTD to switch to another output destination for logging. Initially, there are no devices available. So the only destination of log messages is system console (via `early_print`). But as the kernel above ostd starts taking control and enabling more devices. The kernel should be able to switch to another output destination. For example, Linux uses a ring buffer as a buffer for kernel log messages and a user-space daemon to consume the content in the ring buffer. This approach is more efficient and powerful than doing it with the system console.